fix(cargo_provider): support workspace virtual manifests

jcfangc opened the pull request:
## Description

Fix `cargo_provider` so `cz bump` works with Cargo **workspace virtual manifests** (root `Cargo.toml` containing `[workspace]` / `[workspace.package]` but no `[package]`) when workspace members inherit the version via `package.version.workspace = true`.

Previously, `cz bump` could crash in `commitizen/providers/cargo_provider.py` with:

* `NonExistentKey: Key "package" does not exist` (virtual manifest has no `[package]`)
* `TypeError: expected root version to be a string` / unsafe assumptions around `package.version` shape

This change makes the provider:

* Read/write the version from `[workspace.package].version` when present (fallback to `[package].version` for non-workspace manifests).
* When updating `Cargo.lock`:

  * For non-workspace manifests, update the root package entry.
  * For workspace virtual manifests, scan `workspace.members` (respecting `exclude`) and update lock entries for crates that declare `package.version.workspace = true`.

Fixes #1732

---

## Checklist

* [x] I have read the contributing guidelines
* [x] Added test coverage for the introduced behavior and edge cases
* [x] Ran the test suite locally
* [x] Manually tested the behavior (workspace virtual manifest + non-workspace crate)
* [ ] Updated documentation (not needed for this change)

> Note: the project uses `uv` / `poe` tasks (not Poetry). I ran the equivalent `poe` / `uv`-based test commands.

---

## Expected Behavior

`cz bump` with `version_provider = "cargo"` should succeed for Cargo workspaces using a virtual manifest and centralized versioning in `[workspace.package].version`, without crashing. The workspace version should be updated, and `Cargo.lock` entries for crates inheriting the workspace version (`package.version.workspace = true`) should be updated safely.

Non-workspace projects (root `[package]`) keep existing behavior.

---

## Steps to Test This Pull Request

### Manual test (workspace virtual manifest)

1. Create a workspace root `Cargo.toml`:

   ```toml
   [workspace]
   members = ["crates/*"]

   [workspace.package]
   version = "1.0.0"
   edition = "2024"
   rust-version = "1.92"
   repository = "https://example.com/repo.git"
   ```

2. Add a member crate `crates/common-core/Cargo.toml`:

   ```toml
   [package]
   name = "common-core"
   version.workspace = true
   edition.workspace = true
   rust-version.workspace = true
   repository.workspace = true
   ```

3. Generate `Cargo.lock` once:

   ```bash
   cargo build
   ```

4. Configure Commitizen:

   ```toml
   [tool.commitizen]
   name = "cz_conventional_commits"
   tag_format = "v$version"
   version_scheme = "semver"
   version_provider = "cargo"
   update_changelog_on_bump = true
   version_files = ["Cargo.toml:workspace.package.version"]
   ```

5. Run:

   ```bash
   cz bump --yes
   ```

6. Verify:

   * `[workspace.package].version` is updated (e.g., `1.0.0 -> 1.1.0`)
   * `Cargo.lock` entries for workspace crates inheriting the version are updated
   * no exceptions are raised

### Manual test (non-workspace manifest)

1. Use a standard crate with root `[package]` and an existing `Cargo.lock`.
2. Run `cz bump --yes` and verify the root package version in `Cargo.lock` is updated as before.

---

## Additional Context

* Reported in #1732.
* Targets workspace virtual manifests (no root `[package]`) and avoids unsafe assumptions about the TOML shape of version fields.

jcfangc commented:
> Hi @jcfangc , Thank you for your contribution! Please follow the contribution guideline. You forget to run `poetry format` on the latest iteration.

Sorry about that—I’ll run poetry format and push an update asap.

noirbizarre commented:
I think it would be easier and cleaner to just rebase/squash on the current master.
You are not changing dependencies, so your PR should not touch the `uv.lock` file.
To avoid this side effect, run `uv sync --frozen --all-groups` instead of just `uv sync --all-groups` (`uv` default behavior is to check and update the lock).
Also, you might want to update your `uv` version, it is trying to downgrade the lock version so it might be outdated.

jcfangc commented:
> I think it would be easier and cleaner to just rebase/squash on the current master. You are not changing dependencies, so your PR should not touch the `uv.lock` file. To avoid this side effect, run `uv sync --frozen --all-groups` instead of just `uv sync --all-groups` (`uv` default behavior is to check and update the lock). Also, you might want to update your `uv` version, it is trying to downgrade the lock version so it might be outdated.

Thanks for the feedback.

I’ve rebased/squashed the branch on top of the current master and ensured the PR no longer touches lockfiles:
- reverted/restored `uv.lock` and `poetry.lock` to match `master` (no lockfile changes in the PR diff now)

Going forward I’ll use `uv sync --frozen --all-groups` to avoid accidental lock updates, and I’ll also update my local `uv` to prevent lock format/version downgrades.

All checks are now green.

woile commented on commitizen/providers/cargo_provider.py:
```diff
@@ -3,103 +3,128 @@
+            yield Path(p)
+
+
+def _workspace_inheriting_member_names(
```
what is this supposed to do?

woile commented on commitizen/providers/cargo_provider.py:
```diff
@@ -3,103 +3,128 @@
+    return pkg
+
+
+def _is_workspace_inherited_version(v: Any) -> bool:
```
what is this supposed to do?

Lee-W commented on commitizen/providers/cargo_provider.py:
```diff
@@ -50,56 +52,134 @@ def set_lock_version(self, version: str) -> None:
+            yield Path(match)
+
+
+def _workspace_inheriting_member_names(
```
```suggestion
def _collect_workspace_version_inheriting_members(
```

Lee-W reviewed (commented):
a few nits, but high-level idea looks good

bearomorphism commented:
conflict. converting this PR to draft
